File: java/chatbot/src/test/java/com/rbleggi/chatbot/ChatbotTest.java
package com.rbleggi.chatbot;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

class ChatbotTest {
    private ChatbotSystem keywordBot;
    private ChatbotSystem patternBot;
    private ChatbotSystem intentBot;

    @BeforeEach
    void setUp() {
        keywordBot = new ChatbotSystem(new KeywordMatchingStrategy());
        patternBot = new ChatbotSystem(new PatternMatchingStrategy());
        intentBot = new ChatbotSystem(new IntentBasedStrategy());
    }

    @Test
    @DisplayName("KeywordMatchingStrategy recognizes greeting")
    void keywordStrategy_recognizesGreeting() {
        ChatResponse response = keywordBot.sendMessage("Ola");

        assertTrue(response.text().contains("Ola"));
        assertTrue(response.confidence() > 0.80);
        assertEquals("Keyword-matching", response.strategy());
    }

    @Test
    @DisplayName("KeywordMatchingStrategy handles unknown keywords")
    void keywordStrategy_handlesUnknownKeywords() {
        ChatResponse response = keywordBot.sendMessage("xyz abc 123");

        assertTrue(response.text().contains("nao entendi"));
        assertTrue(response.confidence() < 0.50);
    }

    @Test
    @DisplayName("PatternMatchingStrategy matches price queries")
    void patternStrategy_matchesPriceQueries() {
        ChatResponse response = patternBot.sendMessage("Qual o preco do produto?");

        assertTrue(response.text().toLowerCase().contains("preco") || response.text().toLowerCase().contains("site"));
        assertTrue(response.confidence() > 0.85);
        assertEquals("Pattern-matching", response.strategy());
    }

    @Test
    @DisplayName("PatternMatchingStrategy matches delivery queries")
    void patternStrategy_matchesDeliveryQueries() {
        ChatResponse response = patternBot.sendMessage("Quando chega a entrega?");

        assertTrue(response.text().contains("prazo"));
        assertTrue(response.confidence() > 0.85);
    }

    @Test
    @DisplayName("IntentBasedStrategy detects greeting intent")
    void intentStrategy_detectsGreetingIntent() {
        ChatResponse response = intentBot.sendMessage("Bom dia");

        assertTrue(response.text().contains("bem-vindo") || response.text().contains("Ola"));
        assertTrue(response.confidence() >= 0.70);
        assertEquals("Intent-based", response.strategy());
    }

    @Test
    @DisplayName("IntentBasedStrategy detects payment intent")
    void intentStrategy_detectsPaymentIntent() {
        ChatResponse response = intentBot.sendMessage("Aceita cartao de credito e PIX?");

        assertTrue(response.text().contains("pagamento") || response.text().contains("cartao"));
        assertTrue(response.confidence() >= 0.70);
    }

    @Test
    @DisplayName("ChatbotSystem maintains conversation history")
    void chatbotSystem_maintainsHistory() {
        keywordBot.sendMessage("Ola");
        keywordBot.sendMessage("Qual o preco?");

        List<Message> history = keywordBot.getHistory();
        assertTrue(history.size() >= 2);
    }

    @Test
    @DisplayName("ChatbotSystem can clear history")
    void chatbotSystem_clearsHistory() {
        keywordBot.sendMessage("Ola");
        keywordBot.clearHistory();

        List<Message> history = keywordBot.getHistory();
        assertEquals(0, history.size());
    }

    @Test
    @DisplayName("Different strategies produce different confidence levels")
    void differentStrategies_produceDifferentConfidence() {
        String message = "Ola, qual o preco?";

        ChatResponse r1 = keywordBot.sendMessage(message);
        ChatResponse r2 = patternBot.sendMessage(message);
        ChatResponse r3 = intentBot.sendMessage(message);

        assertNotEquals(r1.strategy(), r2.strategy());
        assertNotEquals(r2.strategy(), r3.strategy());
    }
}

File: kotlin/chatbot/src/test/kotlin/com/rbleggi/chatbot/ChatbotTest.kt
package com.rbleggi.chatbot

import kotlin.test.*

class ChatbotTest {

    private val testMessage = Message("Joao", "Ola, bom dia!")

    @Test
    fun `KeywordMatchingStrategy responds to keywords`() {
        val strategy = KeywordMatchingStrategy()
        val response = strategy.respond(Message("User", "Qual o preco?"))

        assertTrue(response.response.contains("preco") || response.response.contains("R$"))
        assertTrue(response.confidence > 0.5)
    }

    @Test
    fun `KeywordMatchingStrategy handles unknown input`() {
        val strategy = KeywordMatchingStrategy()
        val response = strategy.respond(Message("User", "xyzabc"))

        assertTrue(response.response.contains("Desculpe") || response.response.contains("nao entendi"))
        assertTrue(response.confidence < 0.5)
    }

    @Test
    fun `PatternMatchingStrategy matches greeting pattern`() {
        val strategy = PatternMatchingStrategy()
        val response = strategy.respond(testMessage)

        assertTrue(response.response.contains("Ola") || response.response.contains("ajudar"))
        assertTrue(response.confidence > 0.8)
    }

    @Test
    fun `PatternMatchingStrategy matches pricing pattern`() {
        val strategy = PatternMatchingStrategy()
        val response = strategy.respond(Message("User", "Quanto custa o produto?"))

        assertTrue(response.response.contains("R$") || response.response.contains("valor"))
    }

    @Test
    fun `IntentBasedStrategy identifies greeting intent`() {
        val strategy = IntentBasedStrategy()
        val response = strategy.respond(testMessage)

        assertTrue(response.strategy.contains("Intent-Based"))
        assertTrue(response.confidence > 0.0)
    }

    @Test
    fun `IntentBasedStrategy identifies pricing intent`() {
        val strategy = IntentBasedStrategy()
        val response = strategy.respond(Message("User", "Qual o preco?"))

        assertTrue(response.response.contains("R$") || response.response.contains("produto"))
    }

    @Test
    fun `IntentBasedStrategy handles unknown intent`() {
        val strategy = IntentBasedStrategy()
        val response = strategy.respond(Message("User", "blablabla xyzabc"))

        assertTrue(response.strategy.contains("unknown"))
        assertTrue(response.confidence < 0.5)
    }

    @Test
    fun `Chatbot stores conversation history`() {
        val bot = Chatbot(KeywordMatchingStrategy())
        bot.chat(testMessage)
        bot.chat(testMessage)

        assertEquals(2, bot.getHistory().size)
    }

    @Test
    fun `Chatbot clears history`() {
        val bot = Chatbot(KeywordMatchingStrategy())
        bot.chat(testMessage)
        bot.clearHistory()

        assertEquals(0, bot.getHistory().size)
    }

    @Test
    fun `Chatbot returns response from strategy`() {
        val bot = Chatbot(KeywordMatchingStrategy())
        val response = bot.chat(Message("User", "ola"))

        assertNotNull(response.response)
        assertTrue(response.confidence >= 0.0)
    }

    @Test
    fun `Message data class stores fields correctly`() {
        val message = Message("User", "Text")

        assertEquals("User", message.user)
        assertEquals("Text", message.text)
    }

    @Test
    fun `ChatResponse contains all fields`() {
        val response = ChatResponse("response", 0.9, "strategy")

        assertEquals("response", response.response)
        assertEquals(0.9, response.confidence)
        assertEquals("strategy", response.strategy)
    }
}
